Add px5g WolfSSL version

aparcar opened the pull request:
With the recent choice for Wolfssl as default crypto library the `px5g` tool should also rely on that library instead of `mbedtls`.

This is an adventurous reimplementation of the existing `px5g` tools and examples [found at WolfSSL](https://github.com/wolfSSL/wolfssl-examples).

This version does not yet support subject settings nor EC key options. The former because I couldn't find a WolfSSL function that conveniently parses a single subject argument, the latter because I wasn't sure if these options are even of interest. The original `px5g` uses an RSA key length of 512 per default, so I figured the curves are maybe outdated as well?

@zx2c4 as discussed via chat, feel free to toast this.

aparcar committed:
px5g-wolfssl: add package

This package creates certificates and private keys, just like `px5g`
does. Hower it uses WolfSSL rather than MbedTLS.

Signed-off-by: Paul Spooren <mail@aparcar.org>

aparcar committed:
wolfssl: add certgen config option

The option allows to generate certificates.

Signed-off-by: Paul Spooren <mail@aparcar.org>

aparcar commented:
@dangowrt @jow- please have a look as well

dangowrt reviewed (changes requested):
LGTM apart from not handling errors of `fopen` (EPERM, ENOENT, ...) inhertited from px5g-mbedtls.

dangowrt commented on package/libs/wolfssl/Config.in:
```diff
@@ -16,6 +16,10 @@ config WOLFSSL_HAS_ARC4
 
+config WOLFSSL_HAS_CERTGEN
+	bool "Include certificate generation support"
+	default n
```
I assume we'd need to set this to `y` in order for px5g-wolfssl to work.
What's the size increase of libwolfssl when doing so (ie. is it less than px5g-standalone...)?

aparcar replied:
The option is automatically selected if pg5x-wolfssl is selected.

dangowrt replied:
I don't think this will work on buildbots, we don't have a way for packages to select options of other packages. Please make it default to 'y'.

ynezz commented on package/utils/px5g-wolfssl/px5g-wolfssl.c:
```diff
@@ -0,0 +1,301 @@
+    byte pemBuf[FOURK_SZ];
+    int pemSz = -1;
+    int derSz = -1;
+    wc_InitCert(&newCert);
```
Missing error handling.

aparcar replied:
Actually this was redundant with another init a few lines later.

ynezz commented on package/utils/px5g-wolfssl/px5g-wolfssl.c:
```diff
@@ -0,0 +1,301 @@
+    memset(pemBuf, 0, sizeof(pemBuf));
+    ret = wc_InitCert(&newCert);
+    if (ret != 0) {
+	printf("Init Cert failed: %d\n", ret);
```
You shouldn't continue here.

aparcar replied:
Added return

ynezz commented on package/utils/px5g-wolfssl/px5g-wolfssl.c:
```diff
@@ -0,0 +1,301 @@
+	ret = wc_MakeCert(&newCert, derBuf, sizeof(derBuf), &rsaKey, NULL, rng);
+    }
+    if (ret <= 0) {
+	printf("Make Cert failed: %d\n", ret);
```
You shouldn't continue here.

aparcar replied:
Added return

ynezz commented on package/utils/px5g-wolfssl/px5g-wolfssl.c:
```diff
@@ -0,0 +1,301 @@
+	    NULL, rng);
+    }
+    if (ret <= 0) {
+	printf("Sign Cert failed: %d\n", ret);
```
You shouldn't continue here.

aparcar replied:
Added return

ynezz commented on package/utils/px5g-wolfssl/px5g-wolfssl.c:
```diff
@@ -0,0 +1,301 @@
+
+    ret = wc_DerToPem(derBuf, derSz, pemBuf, sizeof(pemBuf), CERTREQ_TYPE);
+    if (ret <= 0) {
+	printf("DER to PEM failed: %d\n", ret);
```
You shouldn't continue here.

aparcar replied:
Added return

ynezz commented on package/utils/px5g-wolfssl/px5g-wolfssl.c:
```diff
@@ -0,0 +1,301 @@
+    }
+    pemSz = ret;
+
+    write_file(pemBuf, pemSz, certpath);
```
No error checking.

aparcar replied:
Error checking added

ynezz commented on package/utils/px5g-wolfssl/px5g-wolfssl.c:
```diff
@@ -0,0 +1,301 @@
+	(void)ecKey; /* Not using ecKey in RSA_KEY_TYPE case */
+    }
+    if (ret != 0) {
+	printf("Key initialization failed: %d\n", ret);
```
You shouldn't continue here.

aparcar replied:
Added return

ynezz commented on package/utils/px5g-wolfssl/px5g-wolfssl.c:
```diff
@@ -0,0 +1,301 @@
+    if (ret != 0)
+	return ret;
+
+    ret = write_key(&ecKey, &rsaKey, type, keySz, path, pem);
```
Missing error handling.

ynezz commented on package/utils/px5g-wolfssl/px5g-wolfssl.c:
```diff
@@ -0,0 +1,301 @@
+    WC_RNG rng;
+    ret = wc_InitRng(&rng);
+    if (ret != 0)
+	return ret;
```
Probably some error message here?

aparcar replied:
Error message added

ynezz commented on package/utils/px5g-wolfssl/Makefile:
```diff
@@ -0,0 +1,54 @@
+
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=px5g-wolfssl
```
Is there any specific reason for complete new package and not adding this into existing `px5g` package?

dangowrt replied:
Imho it's nicer than combining this into one package `px5g`, as in that way we don't need to filter build-dependencies and yet won't require users of either package to build both mbedtls and wolfssl as build-dependencies of the combined `px5g` package. px5g-wolfssl anyway doesn't share any code with px5g-mbedtls.

aparcar replied:
I agree with @dango, the codebase is entirely different

ynezz commented:
FYI scan-build returns following warnings:

```
px5g-wolfssl.c:157:7: warning: Potential leak of memory pointed to by 'newc'
px5g-wolfssl.c:207:5: warning: Value stored to 'derSz' is never read
```

aparcar commented:
> px5g-wolfssl.c:157:7: warning: Potential leak of memory pointed to by 'newc'
px5g-wolfssl.c:207:5: warning: Value stored to 'derSz' is never read

Thanks. I removed the `subject` handling for now as I have no elegant parsing idea.

aparcar commented on package/utils/px5g-wolfssl/px5g-wolfssl.c:
```diff
@@ -0,0 +1,301 @@
+    int ret;
+    FILE* file;
+    if (path) {
+	file = fopen(path, "wb");
```
Added an error check.

aparcar commented:
Looking at https://safecurves.cr.yp.to/it seems as we shouldn't use any of the NIST curves and therefore don't require the parameter?

zx2c4 commented:
Do you actually need all these options? Why not have this tool just generate one type of certificate always? For that, since this is for webpages, probably the P256 curve for ECDSA makes the most sense for openwrt, especially given how slow RSA is.

neheb commented:
> Looking at https://safecurves.cr.yp.to/it seems as we shouldn't use any of the NIST curves and therefore don't require the parameter?

That website is misleading. It's not that P256 is weak, it's that its creation was not done properly. It's not as ideal of a curve as some of the others on that list, but it's absolutely fine.

zx2c4 commented:
> > Looking at https://safecurves.cr.yp.to/it seems as we shouldn't use any of the NIST curves and therefore don't require the parameter?
> 
> That website is misleading. It's not that P256 is weak, it's that its creation was not done properly. It's not as ideal of a curve as some of the others on that list, but it's absolutely fine.

You're both not really approaching this right.

We're making x509 certificates for webpages right? If that's the case, Ed25519 and such isn't really feasible. P256 is what makes the most sense in this context.

neheb commented:
> > > Looking at https://safecurves.cr.yp.to/it seems as we shouldn't use any of the NIST curves and therefore don't require the parameter?
> > 
> > 
> > That website is misleading. It's not that P256 is weak, it's that its creation was not done properly. It's not as ideal of a curve as some of the others on that list, but it's absolutely fine.
> 
> You're both not really approaching this right.
> 
> We're making x509 certificates for webpages right? If that's the case, Ed25519 and such isn't really feasible. P256 is what makes the most sense in this context.

Correct.  My comment was more geared towards the IRC comment that the curve is "outdated",

I was looking at ssllabs to see which clients would break. Turns out, not many. Even Android 4.4 supports ECDSA.

neheb commented on package/utils/px5g-wolfssl/Makefile:
```diff
@@ -0,0 +1,54 @@
+
+PKG_NAME:=px5g-wolfssl
+PKG_RELEASE:=1
+PKG_LICENSE:=LGPL-2.1
```
erm LGPL is for libraries. This is a utility.

aparcar replied:
Updated to GPL-2.0-or-later

neheb commented on package/utils/px5g-wolfssl/Makefile:
```diff
@@ -0,0 +1,54 @@
+TARGET_LDFLAGS += -lwolfssl
+
+
+TARGET_CFLAGS += -Wl,--gc-sections -Wall -Werror
```
Wall and Werror looks like a pointless development artifact.

aparcar replied:
Removed, these were copied from the original px5g.

neheb commented on package/utils/px5g-wolfssl/px5g-wolfssl.c:
```diff
@@ -0,0 +1,309 @@
+	return ret;
+    }
+
+    if (!argv[1]) {
```
I don't understand this one. This if is blank. Also, the other if statements are just return statements. There's no point in using else after return.

aparcar replied:
I changed the logic a bit

aparcar commented:
I updated the script:
* Use `strncmp` everywhere instead of `strcmp` like the original script
* Use ECC_SECP256R1 per default instead of RSA [which is NIST P-256](https://tools.ietf.org/search/rfc4492#appendix-A)
* Implement `-keyopt` which support P-256, P-384 and P-521
* Set certificate days to 10 years (3650) as requested by @hauke 

This is the first time I write a C tool (except for calculators in some university courses), none of what you see is based on *my style*, please feel free to nitpick on anything.

neheb commented:
I would run this under clang-tidy. Maybe even clang-format.

aparcar commented:
> I would run this under clang-tidy. Maybe even clang-format.

I ran clang-format 10 which changes the style a bit.

clang-tidy gives me some warnings which criticize WolfSSL functions, not sure what I can do about it

```
user@dawn:~/src/openwrt/openwrt/package/utils/px5g-wolfssl$ clang-tidy px5g-wolfssl.c 
Error while trying to load a compilation database:
Could not auto-detect compilation database for file "px5g-wolfssl.c"
No compilation database found in /home/user/src/openwrt/openwrt/package/utils/px5g-wolfssl or any parent directory
fixed-compilation-database: Error while opening fixed database: No such file or directory
json-compilation-database: Error while opening JSON database: No such file or directory
Running without flags.
4 warnings generated.
/home/user/src/openwrt/openwrt/package/utils/px5g-wolfssl/px5g-wolfssl.c:48:5: warning: Call to function 'memset' is insecure as it does not provide security checks introduced in the C11 standard. Replace with analogous functions that support length arguments or provides boundary checks such as 'memset_s' in case of C11 [clang-analyzer-security.insecureAPI.DeprecatedOrUnsafeBufferHandling]
    XMEMSET(der, 0, FOURK_SZ);
    ^
/usr/local/include/wolfssl/wolfcrypt/types.h:448:35: note: expanded from macro 'XMEMSET'
        #define XMEMSET(b,c,l)    memset((b),(c),(l))
                                  ^
/home/user/src/openwrt/openwrt/package/utils/px5g-wolfssl/px5g-wolfssl.c:48:5: note: Call to function 'memset' is insecure as it does not provide security checks introduced in the C11 standard. Replace with analogous functions that support length arguments or provides boundary checks such as 'memset_s' in case of C11
    XMEMSET(der, 0, FOURK_SZ);
    ^
/usr/local/include/wolfssl/wolfcrypt/types.h:448:35: note: expanded from macro 'XMEMSET'
        #define XMEMSET(b,c,l)    memset((b),(c),(l))
                                  ^
/home/user/src/openwrt/openwrt/package/utils/px5g-wolfssl/px5g-wolfssl.c:49:5: warning: Call to function 'memset' is insecure as it does not provide security checks introduced in the C11 standard. Replace with analogous functions that support length arguments or provides boundary checks such as 'memset_s' in case of C11 [clang-analyzer-security.insecureAPI.DeprecatedOrUnsafeBufferHandling]
    XMEMSET(pem, 0, FOURK_SZ);
    ^
/usr/local/include/wolfssl/wolfcrypt/types.h:448:35: note: expanded from macro 'XMEMSET'
        #define XMEMSET(b,c,l)    memset((b),(c),(l))
                                  ^
/home/user/src/openwrt/openwrt/package/utils/px5g-wolfssl/px5g-wolfssl.c:49:5: note: Call to function 'memset' is insecure as it does not provide security checks introduced in the C11 standard. Replace with analogous functions that support length arguments or provides boundary checks such as 'memset_s' in case of C11
    XMEMSET(pem, 0, FOURK_SZ);
    ^
/usr/local/include/wolfssl/wolfcrypt/types.h:448:35: note: expanded from macro 'XMEMSET'
        #define XMEMSET(b,c,l)    memset((b),(c),(l))
                                  ^
/home/user/src/openwrt/openwrt/package/utils/px5g-wolfssl/px5g-wolfssl.c:61:2: warning: Call to function 'memset' is insecure as it does not provide security checks introduced in the C11 standard. Replace with analogous functions that support length arguments or provides boundary checks such as 'memset_s' in case of C11 [clang-analyzer-security.insecureAPI.DeprecatedOrUnsafeBufferHandling]
        memset(pem, 0, sizeof(pem));
        ^
/home/user/src/openwrt/openwrt/package/utils/px5g-wolfssl/px5g-wolfssl.c:61:2: note: Call to function 'memset' is insecure as it does not provide security checks introduced in the C11 standard. Replace with analogous functions that support length arguments or provides boundary checks such as 'memset_s' in case of C11
/home/user/src/openwrt/openwrt/package/utils/px5g-wolfssl/px5g-wolfssl.c:185:5: warning: Call to function 'memset' is insecure as it does not provide security checks introduced in the C11 standard. Replace with analogous functions that support length arguments or provides boundary checks such as 'memset_s' in case of C11 [clang-analyzer-security.insecureAPI.DeprecatedOrUnsafeBufferHandling]
    memset(pemBuf, 0, sizeof(pemBuf));
    ^
/home/user/src/openwrt/openwrt/package/utils/px5g-wolfssl/px5g-wolfssl.c:185:5: note: Call to function 'memset' is insecure as it does not provide security checks introduced in the C11 standard. Replace with analogous functions that support length arguments or provides boundary checks such as 'memset_s' in case of C11

```

neheb commented:
Use -checks=\*

I would ignore the memset warnings. They could be avoided by doing variable = {};

neheb commented on package/utils/px5g-wolfssl/px5g-wolfssl.c:
```diff
@@ -0,0 +1,312 @@
+  char *subject = "";
+  int keySz = 2048;
+  int type = EC_KEY_TYPE;
+  int exp = 65537;
```
65537 can be replaced with WC_RSA_EXPONENT

aparcar replied:
Done

neheb commented on package/utils/px5g-wolfssl/px5g-wolfssl.c:
```diff
@@ -0,0 +1,312 @@
+  RsaKey rsaKey;
+  int ret;
+  int curve = ECC_SECP256R1;
+  int keySz = 2048;
```
*maybe* change to WOLFSSL_MIN_RSA_BITS . It's kind of pushing it.

aparcar replied:
Shouldn't we use 4096 when we already use a validity of 10 years?

neheb replied:
No. Not only is 4096 slow, it doesn't buy anything. 2048 will not be broken in 10 years.

The difference between RSA and EC is vulnerability to index calculus. If that were not the case, RSA256 == EC256. But since it is not, RSA needs to use ridiculous key lengths. EC256 == RSA3072 ignoring implementation problems and whatever else.

Actually now that I think about it, using WOLFSSL_MIN_RSA_BITS is perfect. In the unlikely even that RSA 2048 becomes insecure, px5g would be automatically migrated.

aparcar replied:
Fine for me. Are you okay with Ecdsa as default?

neheb replied:
I am. Although I don't maintain uhttpd.

aparcar replied:
WOLFSSL_MIN_RSA_BITS doesn't seem to be available, it's only defined in wolfssl/internal.h.

neheb replied:
Ah I missed that.

aparcar commented:
Now supports `-subj` parameter (thanks @dangowrt) just like the original script does. `EMAIL` doesn't seem to be an official things in OpenSSL but the current implementation should work fine.

The `strdup` is maybe a bit overkill to show the user some debugging output, just like the time valid string. The former requires extra memory and the latter needs currently fixing to last longer than 2039. And that for a function which is automatically called n first boot and thereby invisible 99% of the time.

neheb commented on package/utils/px5g-wolfssl/px5g-wolfssl.c:
```diff
@@ -0,0 +1,346 @@
+  int type = EC_KEY_TYPE;
+  int exp = WC_RSA_EXPONENT;
+  int curve = ECC_SECP256R1;
+  unsigned int days = 3650;
```
Small nitpick, but 3650 days is not 10 years. That's 3652.5. https://duckduckgo.com/?q=10+years+in+days&ia=answer

aparcar replied:
Fixed, thanks.

aparcar commented:
@neheb can you help me out do make this thing 2040 proof?

aparcar committed:
px5g: rename to px5g-mbedtls

Two versions of `px5g` exists without sharing code. For clarification
rename the previously existing MbedTLS based version to `px5g-mbedtls`
to exists next to `px5g-wolfssl`.

Rename code file of MbedTLS from `px5g.c` to `px5g-mbedtls.c`.

Signed-off-by: Paul Spooren <mail@aparcar.org>

zx2c4 commented:
@aparcar - my suggestion is that you remove all crypto options and only support P256. Then a lot of the complexity of this goes away.

aparcar commented:
The last commit is only cosmetic to lower the legacy confusion.

aparcar commented:
> @aparcar - my suggestion is that you remove all crypto options and only support P256. Then a lot of the complexity of this goes away.

Fine with me, this was mostly a C tutorial for me anyway :smile: I though to get this merged into 20.x it needs to be a drop in replacement. However if more people agree on P-256 only I'm happy to remove some code!

dangowrt commented on package/utils/px5g-wolfssl/Makefile:
```diff
@@ -0,0 +1,54 @@
+  CATEGORY:=Utilities
+  SUBMENU:=Encryption
+  TITLE:=X.509 certificate generator (using wolfssl)
+  DEPENDS:=+libwolfssl +@CONFIG_WOLFSSL_HAS_CERTGEN
```
This will not work. `+` is not supported for configuration symbols and you have to drop the `CONFIG_` prefix when using the `@` syntax. Make it `@WOLFSSL_HAS_CERTGEN` instead and select that symbol by default in the wolfssl makefile.

aparcar replied:
Done

aparcar commented:
I'd say this is ready for testing. Enabling cergen for wolfssl doesn't seem to have any size impact, I'm guessing it's selected by one of the other options as well. The compiled px5g-wolfssl tool is 5kB in size.

neheb commented on package/utils/px5g-wolfssl/px5g-wolfssl.c:
```diff
@@ -0,0 +1,354 @@
+  int ret;
+  byte der[FOURK_SZ], pem[FOURK_SZ];
+  int derSz, pemSz;
+  XMEMSET(der, 0, FOURK_SZ);
```
Just initialize with {}. No need to call some special macro.

aparcar replied:
Done & works, thanks!

neheb commented on package/utils/px5g-wolfssl/px5g-wolfssl.c:
```diff
@@ -0,0 +1,354 @@
+  derSz = ret;
+
+  if (write_pem) {
+    memset(pem, 0, sizeof(pem));
```
Would pem = {}; work here? Would avoid having to call memset.

aparcar replied:
Done & works, thanks!

neheb commented on package/utils/px5g-wolfssl/px5g-wolfssl.c:
```diff
@@ -0,0 +1,354 @@
+int write_key(ecc_key *ecKey, RsaKey *rsaKey, int type, int keySz, char *fName,
+              bool write_pem) {
+  int ret;
+  byte der[FOURK_SZ], pem[FOURK_SZ];
```
I'd put these on separate lines.

aparcar replied:
Done & works, thanks!

neheb commented on package/utils/px5g-wolfssl/px5g-wolfssl.c:
```diff
@@ -0,0 +1,354 @@
+          " and validity %s-%s\n",
+          subject, fstr, tstr);
+
+  memset(pemBuf, 0, sizeof(pemBuf));
```
pemBuf = {}; would avoid memset.

aparcar replied:
Done & works, thanks!

aparcar commented:
Actually the `strdupa` function causes some trouble...

```
user@dawn:~/src/openwrt/openwrt/package/utils/px5g-wolfssl$ gcc -o wip px5g-wolfssl.c -lwolfssl
px5g-wolfssl.c: In function ‘selfsigned’:
px5g-wolfssl.c:180:17: warning: implicit declaration of function ‘strdupa’; did you mean ‘strdup’? [-Wimplicit-function-declaration]
  180 |       subject = strdupa(arg[1]);
      |                 ^~~~~~~
      |                 strdup
px5g-wolfssl.c:180:15: warning: assignment to ‘char *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
  180 |       subject = strdupa(arg[1]);
      |               ^
/usr/bin/ld: /tmp/ccZbrLTH.o: in function `selfsigned':
px5g-wolfssl.c:(.text+0x984): undefined reference to `strdupa'
collect2: error: ld returned 1 exit status
```

dangowrt commented:
Merged. Thanks!
